File: node/fe/ui/pages/eats.js
'use strict';

/**
 * Page objects for the mini-eats app surfaces. The server renders small,
 * labelled HTML pages -- one per app: the customer's home and restaurant menu,
 * the order tracker, the merchant board, the driver board, the admin overview.
 * Each page object reads by label so the assertions are about the food-delivery
 * platform, not the markup.
 *
 * The base URL is the running service; a page that never loads (service down)
 * surfaces as ScreenNotReady, which the steps turn into Blocked.
 */

const BASE = (process.env.MINI_EATS_URL || 'http://127.0.0.1:8130').replace(/\/+$/, '');

class ScreenNotReady extends Error {}

class EatsPage {
  constructor(page) { this.page = page; this.base = BASE; }

  async open(path) {
    try { await this.page.goto(this.base + path, { waitUntil: 'domcontentloaded', timeout: 15_000 }); }
    catch (err) { throw new ScreenNotReady('mini-eats page ' + path + ' did not load: ' + err.message); }
  }

  async restaurants() {
    await this.page.waitForSelector('ul.restaurants li.restaurant', { timeout: 15_000 }).catch(() => { throw new ScreenNotReady('home never rendered'); });
    return this.page.$$eval('ul.restaurants li.restaurant', (els) => els.map((el) => ({
      id: Number(el.getAttribute('data-id')),
      name: el.querySelector('a') ? el.querySelector('a').textContent.trim() : '',
      href: el.querySelector('a') ? el.querySelector('a').getAttribute('href') : '',
    })));
  }

  async menu() {
    await this.page.waitForSelector('ul.menu li.item', { timeout: 15_000 }).catch(() => { throw new ScreenNotReady('menu never rendered'); });
    return this.page.$$eval('ul.menu li.item', (els) => els.map((el) => ({
      id: Number(el.getAttribute('data-id')),
      name: el.querySelector('.name') ? el.querySelector('.name').textContent.trim() : '',
      priceText: el.querySelector('.price') ? el.querySelector('.price').textContent.trim() : '',
      statusText: el.querySelector('.status') ? el.querySelector('.status').textContent.trim() : '',
    })));
  }

  async order() {
    await this.page.waitForSelector('h1.order-id', { timeout: 15_000 }).catch(() => { throw new ScreenNotReady('order page never rendered'); });
    return { idText: await this.page.$eval('h1.order-id', (e) => e.textContent.trim()), statusText: await this.page.$eval('.status', (e) => e.textContent.trim()), totalText: await this.page.$eval('.total', (e) => e.textContent.trim()) };
  }

  async board() {
    await this.page.waitForSelector('ul.orders', { timeout: 15_000 }).catch(() => { throw new ScreenNotReady('merchant board never rendered'); });
    return this.page.$$eval('ul.orders li.order', (els) => els.map((el) => ({
      id: Number(el.getAttribute('data-id')),
      statusText: el.querySelector('.status') ? el.querySelector('.status').textContent.trim() : '',
      totalText: el.querySelector('.total') ? el.querySelector('.total').textContent.trim() : '',
    })));
  }

  async offers() {
    await this.page.waitForSelector('ul.offers', { timeout: 15_000 }).catch(() => { throw new ScreenNotReady('driver board never rendered'); });
    return this.page.$$eval('ul.offers li.offer', (els) => els.map((el) => ({
      id: Number(el.getAttribute('data-id')),
      totalText: el.querySelector('.total') ? el.querySelector('.total').textContent.trim() : '',
      feeText: el.querySelector('.fee') ? el.querySelector('.fee').textContent.trim() : '',
    })));
  }

  async admin() {
    await this.page.waitForSelector('ul.counts', { timeout: 15_000 }).catch(() => { throw new ScreenNotReady('admin page never rendered'); });
    const revenueText = await this.page.$eval('.revenue', (e) => e.textContent.trim());
    const counts = await this.page.$$eval('ul.counts li.count', (els) => els.map((el) => ({
      status: el.getAttribute('data-status'),
      n: el.querySelector('.n') ? el.querySelector('.n').textContent.trim() : '',
    })));
    return { revenueText, counts };
  }
}

module.exports = { EatsPage, ScreenNotReady, BASE };

File: python/fe/ui/pages/eats.py
"""Page objects for the mini-eats app surfaces. Mirror of node/fe/ui/pages/eats.js.
Reads by label; a page that never loads raises ScreenNotReady (grades Blocked)."""

from __future__ import annotations

import os

BASE = os.environ.get("MINI_EATS_URL", "http://127.0.0.1:8130").rstrip("/")


class ScreenNotReady(Exception):
    pass


class EatsPage:
    def __init__(self, page) -> None:
        self.page = page
        self.base = BASE

    def open(self, path):
        try:
            self.page.goto(self.base + path, wait_until="domcontentloaded", timeout=15000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady(f"mini-eats page {path} did not load: {err}") from err

    def restaurants(self):
        try:
            self.page.wait_for_selector("ul.restaurants li.restaurant", timeout=15000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady("home never rendered") from err
        return self.page.eval_on_selector_all("ul.restaurants li.restaurant", """els => els.map(el => ({
            id: Number(el.getAttribute('data-id')),
            name: el.querySelector('a') ? el.querySelector('a').textContent.trim() : '',
            href: el.querySelector('a') ? el.querySelector('a').getAttribute('href') : '',
        }))""")

    def menu(self):
        try:
            self.page.wait_for_selector("ul.menu li.item", timeout=15000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady("menu never rendered") from err
        return self.page.eval_on_selector_all("ul.menu li.item", """els => els.map(el => ({
            id: Number(el.getAttribute('data-id')),
            name: el.querySelector('.name') ? el.querySelector('.name').textContent.trim() : '',
            priceText: el.querySelector('.price') ? el.querySelector('.price').textContent.trim() : '',
            statusText: el.querySelector('.status') ? el.querySelector('.status').textContent.trim() : '',
        }))""")

    def order(self):
        try:
            self.page.wait_for_selector("h1.order-id", timeout=15000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady("order page never rendered") from err
        return {"idText": self.page.eval_on_selector("h1.order-id", "e => e.textContent.trim()"), "statusText": self.page.eval_on_selector(".status", "e => e.textContent.trim()"), "totalText": self.page.eval_on_selector(".total", "e => e.textContent.trim()")}

    def board(self):
        try:
            self.page.wait_for_selector("ul.orders", timeout=15000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady("merchant board never rendered") from err
        return self.page.eval_on_selector_all("ul.orders li.order", """els => els.map(el => ({
            id: Number(el.getAttribute('data-id')),
            statusText: el.querySelector('.status') ? el.querySelector('.status').textContent.trim() : '',
            totalText: el.querySelector('.total') ? el.querySelector('.total').textContent.trim() : '',
        }))""")

    def offers(self):
        try:
            self.page.wait_for_selector("ul.offers", timeout=15000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady("driver board never rendered") from err
        return self.page.eval_on_selector_all("ul.offers li.offer", """els => els.map(el => ({
            id: Number(el.getAttribute('data-id')),
            totalText: el.querySelector('.total') ? el.querySelector('.total').textContent.trim() : '',
            feeText: el.querySelector('.fee') ? el.querySelector('.fee').textContent.trim() : '',
        }))""")

    def admin(self):
        try:
            self.page.wait_for_selector("ul.counts", timeout=15000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady("admin page never rendered") from err
        revenue_text = self.page.eval_on_selector(".revenue", "e => e.textContent.trim()")
        counts = self.page.eval_on_selector_all("ul.counts li.count", """els => els.map(el => ({
            status: el.getAttribute('data-status'),
            n: el.querySelector('.n') ? el.querySelector('.n').textContent.trim() : '',
        }))""")
        return {"revenueText": revenue_text, "counts": counts}
